.github/workflows/pypi_publish.yml

```diff
@@ -0,0 +1,131 @@
+        shell: bash -l {0}
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Set OpenMP mode
+        if: matrix.openmp == 'on'
+        run: |
+          export MODE="omp"
+          echo "MODE=${MODE}" >> $GITHUB_ENV
+
+      - name: Set MacOS Deployment Target
+        if: runner.os == 'macOS'
+        run: |
+          export MACOSX_DEPLOYMENT_TARGET="10.15"
+          echo "MACOSX_DEPLOYMENT_TARGET=${MACOSX_DEPLOYMENT_TARGET}" >> $GITHUB_ENV
+
+      - uses: conda-incubator/setup-miniconda@v2
+        with:
+          activate-environment: testenv
+          allow-softlinks: true
+          auto-activate-base: false
+          auto-update-conda: true
+          channel-priority: flexible
+          channels: conda-forge
+          miniconda-version: "latest"
+          python-version: ${{ matrix.python-version }}
+          show-channel-urls: true
+          use-only-tar-bz2: false
+
+      - name: Set up Visual Studio shell
+        if: runner.os == 'Windows'
+        uses: egor-tensin/vs-shell@v2
+        with:
+          arch: x64
+
+      - name: Install build requirements with conda
+        if: runner.os != 'Linux'
+        run: |
+          set -vxeuo pipefail
+          conda install -y cmake
+
+      - name: Install OpenMP for macOS with conda
+        if: runner.os == 'macOS' && matrix.openmp == 'on'
+        run: |
+          set -vxeuo pipefail
+          conda install -y -c conda-forge llvm-openmp
+
+      - name: Install mpi4py for macOS with conda
+        if: runner.os == 'macOS'
+        run: |
+          set -vxeuo pipefail
+          conda install -y -c conda-forge mpi4py openmpi
+
+      - name: Install cibuildwheel on Linux
+        if: runner.os == 'Linux'
+        run: |
+          set -vxeuo pipefail
+          python -m pip install --upgrade pip
+          python -m pip install cibuildwheel
+          echo "CIBW_BEFORE_ALL=pip install cmake" >> $GITHUB_ENV
+          PYTH_VER=${{ matrix.python-version }}
+          PYTH_VER=${PYTH_VER//.}
+          echo "CIBW_BUILD="cp$PYTH_VER-manylinux*"" >> $GITHUB_ENV
+
+      - name: Build sdist and wheel and install *.whl on Linux
+        if: runner.os == 'Linux'
+        run: |
+          set -vxeuo pipefail
+          python -VV
+          python -m cibuildwheel --output-dir dist env/python
+          ls -la dist/*
+
+      - name: Build sdist and wheel and install *.whl on Windows/Mac
+        if: runner.os != 'Linux'
+        run: |
+          set -vxeuo pipefail
+          cd env/python
+          python -VV
+          python setup.py sdist bdist_wheel
+          ls -la dist/*
+          mkdir -p ../../dist
+          cp dist/*.whl ../../dist/
+
+      - name: Publish wheels to GitHub artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: wheels
+          path: dist/*.whl
+
+  publish-to-pypi:
+    name: Publish to PyPI
+    needs: build-n-publish
+    runs-on: ubuntu-latest
+    permissions:
+      id-token: write
```

Comment:
I like this approach of managing the upload via a single platform! I believe this will also publish the source distribution (=sdist, tar.gz). I think limiting what job in the matrix of OS & Python versions produces the source distribution can help. Otherwise, we are doing it 4x2 times.